pandas/src/joins_func_helper.pxi.in

```diff
@@ -10,18 +11,25 @@ WARNING: DO NOT edit .pxi FILE directly, .pxi is generated from .pxi.in
 
 {{py:
 
-# table_type, by_dtype
-by_dtypes = [('PyObjectHashTable', 'object'), ('Int64HashTable', 'int64_t')]
+# by_dtype, table_type, init_table, s1, s2, s3, g1, g2
+by_dtypes = [('int64_t', 'Int64HashTable', 'Int64HashTable(right_size)',
+              '.set_item(', ', ', ')',
```

Comment:
The `PyObjectHashTable` turns out to be slower than `dict`; see the _by_object_ test:

```
[  0.00%] · For pandas commit hash c33c4cbe:
[  0.00%] ·· Building for conda-py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt...................................
[  0.00%] ·· Benchmarking conda-py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 10.00%] ··· Running join_merge.merge_asof_by_int.time_merge_asof_by_int                                        23.11ms
[ 20.00%] ··· Running join_merge.merge_asof_by_object.time_merge_asof_by_object                                  25.29ms
[ 30.00%] ··· Running join_merge.merge_asof_int32_noby.time_merge_asof_int32_noby                                13.11ms
[ 40.00%] ··· Running join_merge.merge_asof_multiby.time_merge_asof_multiby                                     470.48ms
[ 50.00%] ··· Running join_merge.merge_asof_noby.time_merge_asof_noby                                             9.54ms
[ 50.00%] · For pandas commit hash 725453de:
[ 50.00%] ·· Building for conda-py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt..................................
[ 50.00%] ·· Benchmarking conda-py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 60.00%] ··· Running join_merge.merge_asof_by_int.time_merge_asof_by_int                                        23.18ms
[ 70.00%] ··· Running join_merge.merge_asof_by_object.time_merge_asof_by_object                                  39.56ms
[ 80.00%] ··· Running join_merge.merge_asof_int32_noby.time_merge_asof_int32_noby                                15.71ms
[ 90.00%] ··· Running join_merge.merge_asof_multiby.time_merge_asof_multiby                                       failed
[100.00%] ··· Running join_merge.merge_asof_noby.time_merge_asof_noby                                             9.38ms
    before     after       ratio
  [725453de] [c33c4cbe]
-   15.71ms    13.11ms      0.83  join_merge.merge_asof_int32_noby.time_merge_asof_int32_noby
-   39.56ms    25.29ms      0.64  join_merge.merge_asof_by_object.time_merge_asof_by_object

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
```

I stumbled on this when I tried the `dict` out of desperation since `PyObjectHashTabe` wasn't able to handle tuples the way I wanted.

I did end-up adding the `stringify()` function, which solved a lot of problems, though has its own significant overheard as evidenced by the spike in runtime for the _multiby_ test. I could switch back to `PyObjectHashTable` since the `stringify()` now simplifies a lot of the code. I just wish there were a fast way to hash numpy arrays.